Retry test on "Address already in use" error

pietern opened the pull request:
Summary: This fixes #13907.

Differential Revision: D13046256

teng-li commented on test/test_c10d.py:
```diff
@@ -72,6 +72,22 @@ def wrapper(*args, **kwargs):
+                if str(error) == "Address already in use":
+                    continue
+                raise
+        raise RuntimeError("Ran out of retry attempts")
```
Can you rephrase the error message to be more specific?

pietern replied:
Refactored to raise the original message.

soumith commented on test/test_c10d.py:
```diff
@@ -72,6 +72,22 @@ def wrapper(*args, **kwargs):
+    def wrapper(*args, **kwargs):
+        for _ in range(10):
+            try:
+                return func(*args, **kwargs)
```
add a small timeout of 100ms or so, otherwise it's probably going to be useless.

pietern replied:
I think it shouldn't matter TBH, because it will try to find a new port every time. But it doesn't hurt and this path should almost never hit anyway, so I'll add a random sleep to add some jitter as well.

teng-li commented:
Also, can you add the same functionality for test_distributed and test_distributed_thd. These are the places where we still test TCPStore, if the retry logic hasn't been added there.

pietern committed:
Retry test on "Address already in use" error (#13911)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13911

This fixes #13907.

Differential Revision: D13046256

fbshipit-source-id: af3f6583de18eb010ed5005d86c4c12a5490334f

pietern commented:
@teng-li We don't test the TCPStore functionality there directly -- it still uses the `TEST_MASTER_PORT` constant when the init method is env, and only then will use the TCPStore. I have not seen this cause problems with this fixed port approach but we can revisit if it shows up as flaky.

teng-li commented:
@pietern I remember seeing them happening randomly before as well since the underlying code is the same, while test_distributed will just use the wrapper (env:// will use TCPStore anyway). Cool, let's revisit if it surfaces again.

pietern commented:
@teng-li The underlying code is the same with the difference that is always uses a static port. This is likely why we haven't seen issues with flakiness on CircleCI recently...
